Public/Src/Engine/Scheduler/PipExecutor.cs

```diff
@@ -2849,25 +2852,40 @@ private static int GetCacheLevelForLocality(PublishedEntryRefLocality locality)
             // Only the CanBeReferencedOrCached output will be saved in metadata.StaticOutputHashes
             // We looped metadata.StaticOutputHashes and meanwhile find the corresponding output in current executing pip.
             FileArtifactWithAttributes attributedOutput;
-            using (var poolStringFileArtifactWithAttributes = Pools.GetStringFileArtifactWithAttributesMap())
+            using (var poolAbsolutePathFileArtifactWithAttributes = Pools.GetAbsolutePathFileArtifactWithAttributesMap())
             {
-                Dictionary<string, FileArtifactWithAttributes> outputs = poolStringFileArtifactWithAttributes.Instance;
-                outputs.AddRange(pip.Outputs.ToDictionary(o => o.Path.ToString(pathTable), o => o));
+                Dictionary<AbsolutePath, FileArtifactWithAttributes> outputs = poolAbsolutePathFileArtifactWithAttributes.Instance;
+                outputs.AddRange(pip.Outputs.Select(o => new KeyValuePair<AbsolutePath, FileArtifactWithAttributes>(o.Path, o)));
+
                 for (int i = 0; i < metadata.StaticOutputHashes.Count; i++)
                 {
+                    // TODO: The code path that returns null looks dubious. Could they ever be reached? Should we write a contract here instead of silently concluding weak fingerprint miss?
+
                     FileMaterializationInfo materializationInfo = metadata.StaticOutputHashes[i].Info.ToFileMaterializationInfo(pathTable);

```

Comment:
>metadata.StaticOutputHashes[i] [](start = 66, length = 30)

Many property deref and then index.

Why not change the for-loop to:

```
foreach (var staticOutputHash in metadata.StaticOutputHashes)
{
    ...
}
```
otherwise at least add: 
```
var staticOutputHash = metadata.StaticOutputHashes[i];